### PR TITLE
Refactor make pages

### PR DESCRIPTION
**Pull request by csadorf**
Refactor the make_pages module and related tests. Minor modification to the `apps_meta.schema.json` file, because it did not fully capture the currently used schema.

Resolves issue #21 .

### CODE REVIEW

File: make_ghpages/make_pages.py [hunk 1]
```diff
@@ -80,158 +79,154 @@ def get_git_branches(git_url):
...
-    print("  - apps_meta.json generated")
+    outfile = BUILD_PATH / 'apps_meta.json'
+    rendered = json.dumps(apps_meta, ensure_ascii=False, indent=2)
+    outfile.write_text(rendered, encoding='utf-8')
+    print(f"  - {outfile.relative_to(BUILD_PATH)}")
 
     # Copy schemas
     print("[schemas/v1]")
-    schemas_outdir = Path(outdir_abs).joinpath('schemas/v1')
+    schemas_outdir = BUILD_PATH.joinpath('schemas', 'v1')
```
Comment (CasperWA) [inline make_ghpages/make_pages.py:210]:
```suggestion
    schemas_outdir = BUILD_PATH / 'schemas' / 'v1'
```

File: make_ghpages/make_pages.py [hunk 2]
```diff
@@ -80,158 +79,154 @@ def get_git_branches(git_url):
...
     schemas_outdir.mkdir(parents=True)
-    for schemafile in Path(pwd).glob('../schemas/*.schema.json'):
-        print(f"  - {schemafile.name}")
+    for schemafile in ROOT.glob('schemas/*.schema.json'):
         shutil.copyfile(schemafile, schemas_outdir.joinpath(schemafile.name))
+        print(f"  - {schemas_outdir.relative_to(BUILD_PATH)}/{schemafile.name}")
```
Comment (CasperWA) [inline make_ghpages/make_pages.py:214]:
```suggestion
        print(f"  - {schemas_outdir.relative_to(BUILD_PATH)} / {schemafile.name}")
```

File: make_ghpages/make_pages.py [hunk 3]
```diff
@@ -80,158 +79,154 @@ def get_git_branches(git_url):
...
 
     # Copy schemas
     print("[schemas/v1]")
-    schemas_outdir = Path(outdir_abs).joinpath('schemas/v1')
+    schemas_outdir = BUILD_PATH.joinpath('schemas', 'v1')
     schemas_outdir.mkdir(parents=True)
-    for schemafile in Path(pwd).glob('../schemas/*.schema.json'):
-        print(f"  - {schemafile.name}")
+    for schemafile in ROOT.glob('schemas/*.schema.json'):
         shutil.copyfile(schemafile, schemas_outdir.joinpath(schemafile.name))
```
Comment (CasperWA) [inline make_ghpages/make_pages.py:213]:
```suggestion
        shutil.copyfile(schemafile, schemas_outdir / schemafile.name)
```

File: make_ghpages/make_pages.py [hunk 4]
```diff
@@ -80,158 +79,154 @@ def get_git_branches(git_url):
...
+        print(f"  - {schemas_outdir.relative_to(BUILD_PATH)}/{schemafile.name}")
+
+
+if __name__ == "__main__":
```
Comment (CasperWA) [inline make_ghpages/make_pages.py:217]: Just to keep to either double or single quotation strings.
```suggestion
if __name__ == '__main__':
```

**Review by CasperWA — CHANGES_REQUESTED** [review]
It's really good that these files gets updated to the latest Python syntax and concepts. I believe the original files here were written for Python 2 ...

Nice with some cacheing, but consider using [`CacheControl`](https://cachecontrol.readthedocs.io/en/latest/index.html) instead to be able to handle cache headers as well, see [the note](https://github.com/reclosedev/requests-cache#note) about this in `requests-cache` as well.

Concerning `requests-mock`, `pytest` already has a way of doing this, namely [`monkeypatch`](https://docs.pytest.org/en/stable/monkeypatch.html).

Since this is mainly a syntax and style update as well, you could consider adding function doc strings and type hinting as well. But I don't think these are strictly necessary in this case, since the code is already quite clear in what's going on.

File: make_ghpages/make_pages.py (hunk 2 above)
Comment (csadorf) [reply make_ghpages/make_pages.py:215]: The two path parts are "manually concatenated here, so that wouldn't make sense.

Comment (CasperWA) [reply make_ghpages/make_pages.py:215]: Ah, you're right. So it could be done within a single "{}", but the existing line is fine.

**Comment by csadorf** [comment]
> Nice with some cacheing, but consider using [`CacheControl`](https://cachecontrol.readthedocs.io/en/latest/index.html) instead to be able to handle cache headers as well, see [the note](https://github.com/reclosedev/requests-cache#note) about this in `requests-cache` as well.
> 
Are you ok with leaving as is to avoid further work?

> Concerning `requests-mock`, `pytest` already has a way of doing this, namely [`monkeypatch`](https://docs.pytest.org/en/stable/monkeypatch.html).

I was aware of that, but I found the `requests-mock` way to inject specifically mocked URLs preferable. The package has native support for pytest. Feel free to create a PR into this change set to replace.

> Since this is mainly a syntax and style update as well, you could consider adding function doc strings and type hinting as well. But I don't think these are strictly necessary in this case, since the code is already quite clear in what's going on.

Would prefer to keep *as is* and move forward.

**Comment by csadorf** [comment]
@CasperWA Thank you for your detailed review and great suggestions. I applied almost all of them and responded to your review comment in my previous comment above.

**Comment by CasperWA** [comment]
> > Nice with some cacheing, but consider using [`CacheControl`](https://cachecontrol.readthedocs.io/en/latest/index.html) instead to be able to handle cache headers as well, see [the note](https://github.com/reclosedev/requests-cache#note) about this in `requests-cache` as well.
> 
> Are you ok with leaving as is to avoid further work?

I would strongly prefer header-aware caching. The change shouldn't be big.

> > Concerning `requests-mock`, `pytest` already has a way of doing this, namely [`monkeypatch`](https://docs.pytest.org/en/stable/monkeypatch.html).
> 
> I was aware of that, but I found the `requests-mock` way to inject specifically mocked URLs preferable. The package has native support for pytest. Feel free to create a PR into this change set to replace.

My concern is adding another dependency when it's essentially unnecessary. Changing it after the fact seems sloppy. Do you think the work needed to use `pytest.monkeypatch` instead is too great, and it's worth using `requests-mock` permanently instead? The examples from the linked documentation of `monkeypatch` is mainly all about mocking exactly the `requests` library, and from experience it seems the code you have already put in should be easily transformed to comply with the `monkeypatch` API.

> > Since this is mainly a syntax and style update as well, you could consider adding function doc strings and type hinting as well. But I don't think these are strictly necessary in this case, since the code is already quite clear in what's going on.
> 
> Would prefer to keep _as is_ and move forward.

Fair enough.

**Comment by csadorf** [comment]
> > > Nice with some cacheing, but consider using [`CacheControl`](https://cachecontrol.readthedocs.io/en/latest/index.html) instead to be able to handle cache headers as well, see [the note](https://github.com/reclosedev/requests-cache#note) about this in `requests-cache` as well.
> > 
> > 
> > Are you ok with leaving as is to avoid further work?
> 
> I would strongly prefer header-aware caching. The change shouldn't be big.
> 

Done.

> > > Concerning `requests-mock`, `pytest` already has a way of doing this, namely [`monkeypatch`](https://docs.pytest.org/en/stable/monkeypatch.html).
> > 
> > 
> > I was aware of that, but I found the `requests-mock` way to inject specifically mocked URLs preferable. The package has native support for pytest. Feel free to create a PR into this change set to replace.
> 
> My concern is adding another dependency when it's essentially unnecessary. Changing it after the fact seems sloppy. Do you think the work needed to use `pytest.monkeypatch` instead is too great, and it's worth using `requests-mock` permanently instead? The examples from the linked documentation of `monkeypatch` is mainly all about mocking exactly the `requests` library, and from experience it seems the code you have already put in should be easily transformed to comply with the `monkeypatch` API.
> 

Unless I'm missing something, implementing the exact same mocking behavior in a similar flexible way appears non-trivial to me. I don't really see the issue with adding this dependency. Since you say that you have experience with this, I invite you to implement the change and create a PR directly into this branch (no change after the fact needed).

**Comment by csadorf** [comment]
@CasperWA This PR is currently stalled because of your change request that I'm not willing to meet.

@yakutovicha Can you weigh in?

**Comment by yakutovicha** [comment]
> @CasperWA This PR is currently stalled because of your change request that I'm not willing to meet.
> 
> @yakutovicha Can you weigh in?

I will, but later today. I have meetings now.

**Comment by yakutovicha** [comment]
Okay. I looked in the issue in more details.

> The examples from the linked documentation of `monkeypatch` is mainly all about mocking exactly the `requests` library, and from experience it seems the code you have already put in should be easily transformed to comply with the `monkeypatch` API.

@CasperWA are you aware of an easy way to mock the request function so that it returns different results depending on its arguments?

I tried to implement that and I quickly realized how complex the task is. First of all, one needs to implement a `my-request-mock` function that would treat different inputs argument differently. Further, one would need to implement a `MockResponse` class that would contain properties like `content`, `text` etcetera. Then one needs to make sure the `my-request-mock` function returns different `MockResponse` object depending on the argument. So my understanding that `requests-mock` exactly takes care of that.

Unless I missed something, I believe the approach taken by @csadorf is the easiest one. I believe that adding a new dependency is the fair price for that. @CasperWA, if you have a better idea of how to implement the same thing nicer and easier using monkeypatch - you are of course very much welcome to do this as a separate PR, as it was mentioned earlier.

I think we should go on and invest our time on other things.

**Review by yakutovicha — APPROVED** [review]

**Pull request merged**